File: scripts/generate_report.py
"""
文章分析报告 Word 文档生成脚本（python-docx）

用法：
  echo '<JSON>' | python generate_report.py
  python generate_report.py --input data.json
  python generate_report.py --input data.json --output report.docx

JSON 输入格式见 SKILL.md

依赖：pip install python-docx
"""

import sys
import json
import argparse

from docx import Document
from docx.shared import Pt, Cm, RGBColor, DXA
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.enum.table import WD_TABLE_ALIGNMENT
from docx.oxml.ns import qn


# ============================================================
# 常量
# ============================================================

HEADING1_COLOR = RGBColor(0x1F, 0x38, 0x64)  # 深蓝
HEADING2_COLOR = RGBColor(0x2E, 0x75, 0xB6)  # 蓝
HEADING3_COLOR = RGBColor(0x37, 0x56, 0x23)  # 深绿
BODY_FONT = 'Arial'
BODY_SIZE = Pt(11)


# ============================================================
# 读取输入
# ============================================================

def load_input():
    parser = argparse.ArgumentParser(description='Generate article analysis report')
    parser.add_argument('--input', help='JSON input file path')
    parser.add_argument('--output', help='Output .docx file path')
    args = parser.parse_args()

    if args.input:
        with open(args.input, 'r', encoding='utf-8') as f:
            data = json.load(f)
    else:
        if sys.stdin.isatty():
            print('用法: echo \'<JSON>\' | python generate_report.py', file=sys.stderr)
            print('      python generate_report.py --input data.json [--output report.docx]', file=sys.stderr)
            sys.exit(1)
        data = json.load(sys.stdin)

    data['outputFile'] = args.output or data.get('outputFile', 'article_analysis_report.docx')
    return data


# ============================================================
# 文档构建
# ============================================================

def set_run_font(run, size=BODY_SIZE, color=None, bold=False):
    """统一设置 run 的字体属性"""
    run.font.name = BODY_FONT
    run.font.size = size
    run.font.bold = bold
    if color:
        run.font.color.rgb = color
    # 设置中文字体（兼容东亚文字）
    run._element.rPr.rFonts.set(qn('w:eastAsia'), BODY_FONT)


def add_paragraph(doc, text, size=BODY_SIZE, color=None, bold=False, alignment=None):
    """添加普通段落"""
    p = doc.add_paragraph()
    if alignment is not None:
        p.alignment = alignment
    run = p.add_run(text)
    set_run_font(run, size, color, bold)
    return p


def add_heading1(doc, text):
    """添加 Heading 1（深蓝）"""
    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = p.add_run(text)
    set_run_font(run, Pt(16), HEADING1_COLOR, bold=True)
    p.paragraph_format.space_before = Pt(15)
    p.paragraph_format.space_after = Pt(8)
    return p


def add_heading2(doc, text):
    """添加 Heading 2（蓝）"""
    p = doc.add_paragraph()
    run = p.add_run(text)
    set_run_font(run, Pt(13), HEADING2_COLOR, bold=True)
    p.paragraph_format.space_before = Pt(10)
    p.paragraph_format.space_after = Pt(5)
    return p


def add_heading3(doc, text):
    """添加 Heading 3（深绿）"""
    p = doc.add_paragraph()
    run = p.add_run(text)
    set_run_font(run, Pt(11), HEADING3_COLOR, bold=True)
    p.paragraph_format.space_before = Pt(8)
    p.paragraph_format.space_after = Pt(4)
    return p


def add_spacer(doc):
    """添加空行"""
    p = doc.add_paragraph()
    run = p.add_run('')
    set_run_font(run, Pt(6))
    p.paragraph_format.space_before = Pt(0)
    p.paragraph_format.space_after = Pt(0)
    return p


def add_bullet(doc, text):
    """添加项目符号条目"""
    p = doc.add_paragraph(style='List Bullet')
    run = p.add_run(text)
    set_run_font(run, BODY_SIZE)
    return p


def add_table(doc, sec):
    """添加表格"""
    columns = sec.get('columns', [])
    rows = sec.get('rows', [])
    if not columns or not rows:
        return

    col_count = len(columns)
    table = doc.add_table(rows=1 + len(rows), cols=col_count)
    table.alignment = WD_TABLE_ALIGNMENT.CENTER

    # 表头
    headers = [c if isinstance(c, str) else c.get('header', '') for c in columns]
    for i, header_text in enumerate(headers):
        cell = table.rows[0].cells[i]
        cell.text = ''
        p = cell.paragraphs[0]
        run = p.add_run(header_text)
        set_run_font(run, BODY_SIZE, bold=True)
        # 浅蓝背景
        shading = cell._element.get_or_add_tcPr()
        shd_elm = shading.makeelement(qn('w:shd'), {
            qn('w:fill'): 'D5E8F0',
            qn('w:val'): 'clear'
        })
        shading.append(shd_elm)

    # 数据行
    for row_idx, row_data in enumerate(rows):
        for col_idx, cell_text in enumerate(row_data):
            text = cell_text if isinstance(cell_text, str) else cell_text.get('text', '')
            cell = table.rows[row_idx + 1].cells[col_idx]
            cell.text = ''
            p = cell.paragraphs[0]
            run = p.add_run(text)
            set_run_font(run, BODY_SIZE)

    add_spacer(doc)


def build_document(data):
    """构建完整文档"""
    doc = Document()

    # 页面设置：A4，边距约 2.1cm
    section = doc.sections[0]
    section.page_width = Cm(21.0)
    section.page_height = Cm(29.7)
    section.top_margin = Cm(2.1)
    section.bottom_margin = Cm(2.1)
    section.left_margin = Cm(2.1)
    section.right_margin = Cm(2.1)

    # 页眉
    header = section.header
    header_para = header.paragraphs[0]
    header_para.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = header_para.add_run(data.get('title', '文章分析报告'))
    set_run_font(run, Pt(9), RGBColor(0x88, 0x88, 0x88))

    # 页脚（页码）
    footer = section.footer
    footer_para = footer.paragraphs[0]
    footer_para.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run_prefix = footer_para.add_run('第 ')
    set_run_font(run_prefix, Pt(9), RGBColor(0x88, 0x88, 0x88))
    # 插入页码字段
    fld_char_begin = footer_para._element.makeelement(qn('w:fldChar'), {qn('w:fldCharType'): 'begin'})
    run_page = footer_para.add_run()
    run_page._element.append(fld_char_begin)
    instr_text = footer_para._element.makeelement(qn('w:instrText'), {})
    instr_text.text = ' PAGE '
    run_page2 = footer_para.add_run()
    run_page2._element.append(instr_text)
    fld_char_end = footer_para._element.makeelement(qn('w:fldChar'), {qn('w:fldCharType'): 'end'})
    run_page3 = footer_para.add_run()
    run_page3._element.append(fld_char_end)
    run_suffix = footer_para.add_run(' 页')
    set_run_font(run_suffix, Pt(9), RGBColor(0x88, 0x88, 0x88))

    # 默认字体
    style = doc.styles['Normal']
    font = style.font
    font.name = BODY_FONT
    font.size = BODY_SIZE
    style.element.rPr.rFonts.set(qn('w:eastAsia'), BODY_FONT)

    # ---- 标题 ----
    add_heading1(doc, data.get('title', '文章分析报告'))
    add_spacer(doc)

    # ---- 来源信息 ----
    if data.get('source'):
        add_paragraph(doc, data['source'], Pt(10), RGBColor(0x66, 0x66, 0x66),
                       alignment=WD_ALIGN_PARAGRAPH.CENTER)
    if data.get('sourceUrl'):
        add_paragraph(doc, f"原文链接：{data['sourceUrl']}", Pt(9),
                       RGBColor(0x88, 0x88, 0x88), alignment=WD_ALIGN_PARAGRAPH.CENTER)
    if data.get('sourceUrls'):
        for s in data['sourceUrls']:
            url_text = f"{s.get('title', '')} - {s.get('url', '')}"
            add_paragraph(doc, url_text, Pt(9), RGBColor(0x88, 0x88, 0x88),
                           alignment=WD_ALIGN_PARAGRAPH.CENTER)
    add_spacer(doc)

    # ---- 各章节 ----
    for sec in data.get('sections', []):
        add_heading2(doc, sec['heading'])

        sec_type = sec.get('type', '')
        if sec_type == 'bullets':
            for item in sec.get('items', []):
                add_bullet(doc, item)
        elif sec_type == 'table':
            add_table(doc, sec)
        elif sec_type == 'paragraphs':
            for sub in sec.get('subsections', []):
                add_heading3(doc, sub['subheading'])
                for para_text in sub.get('paragraphs', []):
                    add_paragraph(doc, para_text)
                add_spacer(doc)

        add_spacer(doc)

    # ---- 结尾 ----
    add_paragraph(doc, '—— 报告结束 ——', Pt(10), RGBColor(0xAA, 0xAA, 0xAA),
                   alignment=WD_ALIGN_PARAGRAPH.CENTER)

    return doc


# ============================================================
# 主入口
# ============================================================

def main():
    data = load_input()
    doc = build_document(data)
    doc.save(data['outputFile'])
    print(f"✅ 文档已生成：{data['outputFile']}")


if __name__ == '__main__':
    main()

File: scripts/generate_report.js
/**
 * 文章分析报告 Word 文档生成脚本（docx-js）
 *
 * 用法：
 *   echo '<JSON>' | node generate_report.js
 *   node generate_report.js --input data.json
 *   node generate_report.js --input data.json --output report.docx
 *
 * JSON 输入格式见 SKILL.md
 *
 * 依赖：npm install docx
 */

const fs = require('fs');
const path = require('path');
const {
  Document, Packer, Paragraph, TextRun,
  HeadingLevel, AlignmentType, LevelFormat,
  BorderStyle, WidthType, ShadingType,
  Table, TableRow, TableCell,
  PageBreak, PageNumber, Header, Footer
} = require('docx');

// ============================================================
// 读取输入
// ============================================================

function loadInput() {
  const args = process.argv.slice(2);
  let inputFile = null;
  let outputFile = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--input' && args[i + 1]) inputFile = args[++i];
    else if (args[i] === '--output' && args[i + 1]) outputFile = args[++i];
  }

  let jsonStr;

  if (inputFile) {
    jsonStr = fs.readFileSync(inputFile, 'utf-8');
  } else {
    // 从 stdin 读取
    const chunks = [];
    const stdin = process.stdin;
    if (stdin.isTTY) {
      console.error('用法: echo \'<JSON>\' | node generate_report.js');
      console.error('      node generate_report.js --input data.json [--output report.docx]');
      process.exit(1);
    }
    jsonStr = fs.readFileSync('/dev/stdin', 'utf-8');
  }

  const data = JSON.parse(jsonStr);
  data.outputFile = outputFile || data.outputFile || 'article_analysis_report.docx';
  return data;
}

// ============================================================
// 文档构建
// ============================================================

const A4_WIDTH = 11906;
const A4_HEIGHT = 16838;
const MARGIN = 1200;
const CONTENT_WIDTH = A4_WIDTH - MARGIN * 2;  // ~9506 DXA

function buildChildren(data) {
  const kids = [];

  // 标题
  kids.push(new Paragraph({
    heading: HeadingLevel.HEADING_1,
    alignment: AlignmentType.CENTER,
    children: [new TextRun(data.title || '文章分析报告')]
  }));
  kids.push(spacer());

  // 来源信息
  if (data.source) {
    kids.push(new Paragraph({
      alignment: AlignmentType.CENTER,
      children: [new TextRun({ text: data.source, size: 20, color: '666666', font: 'Arial' })]
    }));
  }
  if (data.sourceUrl) {
    kids.push(new Paragraph({
      alignment: AlignmentType.CENTER,
      children: [new TextRun({ text: `原文链接：${data.sourceUrl}`, size: 18, color: '888888', font: 'Arial' })]
    }));
  }
  // 多来源链接
  if (data.sourceUrls && data.sourceUrls.length > 0) {
    for (const s of data.sourceUrls) {
      kids.push(new Paragraph({
        alignment: AlignmentType.CENTER,
        children: [new TextRun({ text: `${s.title || ''} - ${s.url}`, size: 18, color: '888888', font: 'Arial' })]
      }));
    }
  }
  kids.push(spacer());

  // 各章节
  for (const sec of (data.sections || [])) {
    // Heading 2
    kids.push(new Paragraph({
      heading: HeadingLevel.HEADING_2,
      children: [new TextRun(sec.heading)]
    }));

    if (sec.type === 'bullets') {
      for (const item of (sec.items || [])) {
        kids.push(bullet(item));
      }
    } else if (sec.type === 'table') {
      kids.push(buildTable(sec));
    } else if (sec.type === 'paragraphs') {
      for (const sub of (sec.subsections || [])) {
        // Heading 3
        kids.push(new Paragraph({
          heading: HeadingLevel.HEADING_3,
          children: [new TextRun(sub.subheading)]
        }));
        for (const para of (sub.paragraphs || [])) {
          kids.push(new Paragraph({
            children: [new TextRun({ text: para, font: 'Arial' })]
          }));
        }
        kids.push(spacer());
      }
    }

    kids.push(spacer());
  }

  // 结尾
  kids.push(new Paragraph({
    alignment: AlignmentType.CENTER,
    children: [new TextRun({ text: '—— 报告结束 ——', size: 20, color: 'AAAAAA', font: 'Arial' })]
  }));

  return kids;
}

function buildDocument(children, title) {
  return new Document({
    styles: {
      default: {
        document: { run: { font: 'Arial', size: 22 } }
      },
      paragraphStyles: [
        {
          id: 'Heading1', name: 'Heading 1', basedOn: 'Normal', next: 'Normal', quickFormat: true,
          run: { size: 32, bold: true, font: 'Arial', color: '1F3864' },
          paragraph: { spacing: { before: 300, after: 150 } }
        },
        {
          id: 'Heading2', name: 'Heading 2', basedOn: 'Normal', next: 'Normal', quickFormat: true,
          run: { size: 26, bold: true, font: 'Arial', color: '2E75B6' },
          paragraph: { spacing: { before: 200, after: 100 } }
        },
        {
          id: 'Heading3', name: 'Heading 3', basedOn: 'Normal', next: 'Normal', quickFormat: true,
          run: { size: 22, bold: true, font: 'Arial', color: '375623' },
          paragraph: { spacing: { before: 150, after: 80 } }
        }
      ]
    },
    numbering: {
      config: [
        {
          reference: 'bullets',
          levels: [{
            level: 0, format: LevelFormat.BULLET, text: '\u2022',
            alignment: AlignmentType.LEFT,
            style: { paragraph: { indent: { left: 720, hanging: 360 } } }
          }]
        }
      ]
    },
    sections: [{
      properties: {
        page: {
          size: { width: A4_WIDTH, height: A4_HEIGHT },
          margin: { top: MARGIN, right: MARGIN, bottom: MARGIN, left: MARGIN }
        }
      },
      headers: {
        default: new Header({
          children: [new Paragraph({
            alignment: AlignmentType.CENTER,
            children: [new TextRun({ text: title || '文章分析报告', size: 18, color: '888888', font: 'Arial' })]
          })]
        })
      },
      footers: {
        default: new Footer({
          children: [new Paragraph({
            alignment: AlignmentType.CENTER,
            children: [
              new TextRun({ text: '第 ', size: 18, color: '888888', font: 'Arial' }),
              new TextRun({ children: [PageNumber.CURRENT], size: 18, color: '888888', font: 'Arial' }),
              new TextRun({ text: ' 页', size: 18, color: '888888', font: 'Arial' })
            ]
          })]
        })
      },
      children
    }]
  });
}

// ---- 辅助函数 ----

function spacer() {
  return new Paragraph({ children: [new TextRun('')] });
}

function bullet(text) {
  return new Paragraph({
    numbering: { reference: 'bullets', level: 0 },
    children: [new TextRun({ text, font: 'Arial' })]
  });
}

function buildTable(sec) {
  const colWidth = Math.floor(CONTENT_WIDTH / sec.columns.length);
  const rows = [
    new TableRow({
      children: sec.columns.map(c => cell(typeof c === 'string' ? c : c.header, true, colWidth))
    }),
    ...sec.rows.map(row =>
      new TableRow({
        children: row.map((c, i) => cell(typeof c === 'string' ? c : c.text, false, colWidth))
      })
    )
  ];
  return new Table({
    width: { size: CONTENT_WIDTH, type: WidthType.DXA },
    columnWidths: sec.columns.map(() => colWidth),
    rows
  });
}

function cell(text, isHeader, width) {
  const border = { style: BorderStyle.SINGLE, size: 1, color: 'CCCCCC' };
  return new TableCell({
    borders: { top: border, bottom: border, left: border, right: border },
    width: { size: width, type: WidthType.DXA },
    shading: isHeader ? { fill: 'D5E8F0', type: ShadingType.CLEAR } : undefined,
    margins: { top: 80, bottom: 80, left: 120, right: 120 },
    children: [new Paragraph({
      children: [new TextRun({ text, bold: isHeader, font: 'Arial' })]
    })]
  });
}

// ============================================================
// 主入口
// ============================================================

function main() {
  const data = loadInput();
  const children = buildChildren(data);
  const doc = buildDocument(children, data.title);

  Packer.toBuffer(doc).then(buffer => {
    fs.writeFileSync(data.outputFile, buffer);
    console.log(`✅ 文档已生成：${data.outputFile}`);
  }).catch(err => {
    console.error('❌ 生成失败：', err);
    process.exit(1);
  });
}

main();
